assets/model_monitoring/components/src/model_data_collector_preprocessor/spark_run.py

```diff
@@ -0,0 +1,169 @@
+        return df  # no dataref column, return df directly
+    if MDC_DATA_COLUMN not in columns:
+        # TODO need full schema override
+        raise ValueError(f"{MDC_DATAREF_COLUMN} column without {MDC_DATA_COLUMN} is not supported yet.")
+
+    @udf(returnType=StringType())
+    def load_json_str(path):
+        return read_json_content(path)
+
+    # TODO separate df into 2 parts, one with dataref, one without, then only handle the one with dataref, and
+    #      union the 2 parts back together
+
+    # load json string from dataref to ref_json_str column
+    df = df.withColumn("ref_json_str", load_json_str(df[MDC_DATAREF_COLUMN]))
+    # parse json string to json object with from_json(), into ref_data column
+    df = df.withColumn("ref_data", from_json(df["ref_json_str"], df.schema[MDC_DATA_COLUMN].dataType))
+    # replace data with ref_data if data column is null
+    df = df.withColumn(MDC_DATA_COLUMN,
+                       when(col(MDC_DATA_COLUMN).isNull(), col("ref_data")).otherwise(col(MDC_DATA_COLUMN)))
+    # drop temp columns
+    df = df.drop("ref_json_str", "ref_data")
+
+    return df
+
+
+def _extract_data_and_correlation_id(df: DataFrame, extract_correlation_id: bool) -> DataFrame:
+    """Extract data and correlation id from the MDC logs."""
+    columns = [MDC_DATA_COLUMN]
+    if extract_correlation_id:
+        columns.append(MDC_CORRELATION_ID_COLUMN)
+        # explode the data column of array type to multiple rows with index
+        df = df[columns].select(posexplode(MDC_DATA_COLUMN).alias("index", "value"), MDC_CORRELATION_ID_COLUMN)
+        # set the new correlationid as {correlationid}_{index}
+        df = df.withColumn(MDC_CORRELATION_ID_COLUMN, concat_ws("_", MDC_CORRELATION_ID_COLUMN, "index")).drop("index")
+        # select the 1st level features as columns
+        df = df.select("value.*", MDC_CORRELATION_ID_COLUMN)
+    else:
+        df = df[columns].select(posexplode(MDC_DATA_COLUMN).alias("index", "value"))
+        df = df.select("value.*")
+    return df
+
+
+def _mdc_uri_folder_to_preprocessed_spark_df(
+        data_window_start: datetime, data_window_end: datetime, input_data: str, extract_correlation_id: bool,
+        add_tags_func=None) -> DataFrame:
+    """Read raw MDC data, preprocess, and return in a Spark DataFrame."""
+    # Parse the dates
+    start_datetime = parser.parse(data_window_start)
+    end_datetime = parser.parse(data_window_end)
+
+    df = _mdc_uri_folder_to_raw_spark_df(start_datetime, end_datetime, input_data, add_tags_func)
+    print("df converted from MDC raw uri folder:")
+    df.select("data").show(truncate=False)
+    df.printSchema()
+
+    df = _load_dataref_into_data_column(df)
+    df.select("data").show(truncate=False)
+    df.printSchema()
+
+    transformed_df = _extract_data_and_correlation_id(df, extract_correlation_id)
+    transformed_df.show()
+    transformed_df.printSchema()
+
+    return transformed_df
```

Comment:
Seems we need to convert the complex type to a string before outputting to mltable.

Do you have a test case to cover that?

Reply:
added UT